### PR TITLE
Upgrade to v2.0

### PR DESCRIPTION
**Pull request by dks17**
In this upgrade all time calculation operations are obtained on MongoDB server, not a client's. Thereby we have no time synchronization problem between clients.

Please, read the docs in this PR for details.

Fixes: #15 #17 #72

**Added**
  * `exponential_backoff` and `locked_at_backoff` methods for timeout calculating between retries to lock a document and `secure_locking_name` method to generate unique name of a lock for `Mongoid::Locker`
  * `::unlock_all` class method to unlock all locked documents in the collection
  * new parameters for `Mongoid::Locker` and a class where it's included:
    - `locking_name_field`
    - `locking_name_length`
    - `lock_timeout`
    - `maximum_backoff`
    - `locker_write_concern`
    - `backoff_algorithm`
      `locking_name_generator`
  * `find_and_lock`, `find_and_unlock`, `locked_at`, and `current_mongodb_time` methods to `Mongoid::Locker::Wrapper`

**Removed**
  * `:timeout` and `:retry_sleep` options from `#with_lock` method
  * `::timeout_lock_after` and `::lock_timeout` class methods
  * `locked_until_field` parameter for `Mongoid::Locker`
  * `update` and `locked_until` methods from `Mongoid::Locker::Wrapper`

To upgrade see [UPGRADING](//github.com/mongoid/mongoid-locker/blob/master/UPGRADING.md) guide.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,88 +28,133 @@ gem 'mongoid-locker'
...
-
-The default timeout can also be set on a per-class basis:
-
-```ruby
-class QueueItem
-  # ...
-  timeout_lock_after 10
-end
-```
+The `#with_lock` function takes an optional [handful of options](https://www.rubydoc.info/gems/mongoid-locker/Mongoid/Locker#with_lock-instance_method), so make sure to take a look.
```
Comment (dblock) [inline README.md]: Minor, we should remove these links, as they point to some latest version of mongoid-locker and should probably be just the list below.

File: README.md [hunk 2]
```diff
@@ -23,88 +28,133 @@ gem 'mongoid-locker'
...
+  locker locked_at_field: :locker_locking_name,
+         locked_at_field: :locker_locked_at
+end
+```
+
+### Available parameters for Mongoid::Locker, a class where it's included
+| parameter | default | options | description |
+|---|---|---|
+| locking_name_field | `:locking_name` | any field name | field where name of locking is storing, must be type of `String` |
+| locked_at_field | `:locked_at` | any field name | field where it is storing the time of beginning a lock of a document, must be type of `Time` |
```
Comment (dblock) [inline README.md]: of type `Time`

File: README.md [hunk 3]
```diff
@@ -23,88 +28,133 @@ gem 'mongoid-locker'
...
+| lock_timeout | `5` | | within this time (in seconds) a document is considering as locked |
```
Comment (dblock) [inline README.md]: considered as

File: README.md [hunk 4]
```diff
@@ -23,88 +28,133 @@ gem 'mongoid-locker'
...
+| locker_write_concern | `{ w: 1 }` | see [MongoDB Write Concern](https://docs.mongodb.com/manual/reference/write-concern/#write-concern-specification)| a write concern only used for lock and unlock operations |
+| locking_name_length | `nil` | accepts `n` option for [SecureRandom.urlsafe_base64](https://ruby-doc.org/stdlib-2.2.3/libdoc/securerandom/rdoc/SecureRandom.html#method-c-urlsafe_base64) | uses by `secure_locking_name` generator |
```
Comment (dblock) [inline README.md]: used by

Comment (dblock) [reply README.md]: Maybe just spell it out? _the length, in bytes, of the random number to be generated_

File: README.md [hunk 5]
```diff
@@ -23,88 +28,133 @@ gem 'mongoid-locker'
...
+| maximum_backoff | `60.0` | | the highest timeout (in seconds) between retires to lock a document, reaching that value `with_lock` method raises [could not get lock error](https://www.rubydoc.info/gems/mongoid-locker/Mongoid/Locker/Errors/DocumentCouldNotGetLock) |
+| backoff_algorithm | `:exponential_backoff` | `:locked_at_backoff` or [custom algorithm](#custom-backoff_algorithm-and-locking_name_generator) | algorithm used for timeout calculating between retries to lock a document|
+| locking_name_generator | `:secure_locking_name` | [custom generator](#custom-backoff_algorithm-and-locking_name_generator) | generator used to generate unique name of a lock |
+
+For instances of a class where `Mongoid::Locker` is included, all parameters of a class available for reading.
```
Comment (dblock) [inline README.md]: are available

File: README.md [hunk 6]
```diff
@@ -23,88 +28,133 @@ gem 'mongoid-locker'
...
+```ruby
+  document.lock_timeout
+  #=> 5
+```
+
+### Custom :backoff_algorithm and :locking_name_generator
+A method which is defined in `Mongoid::Locker` will be available in a class where it is included.
```
Comment (dblock) [inline README.md]: are available

File: README.md [hunk 7]
```diff
@@ -23,88 +28,133 @@ gem 'mongoid-locker'
...
+
+Method `with_lock` passes to the methods a document to which apply `with_lock` and [a hash of options]((https://www.rubydoc.info/gems/mongoid-locker/Mongoid/Locker#with_lock-instance_method). The hash may look like this:
+```ruby
+  {:retries=>Infinity, :attempt=>1, :locking_name=>"71c1ccd4-72d9-4a83-bbed-adf65803bd5d"}
```
Comment (dblock) [inline README.md]: Lets make this modern Ruby like, `retries: Infinity, attempt: 1 ...`

File: README.md [hunk 8]
```diff
@@ -23,88 +28,133 @@ gem 'mongoid-locker'
...
+module Mongoid
+  module Locker
+    def self.custom_backoff(doc, _opts)
+      rand > 0.5 ? 5 : doc.maximum_backoff
+    end
+  end
+end
+```
+
+A custom locking name generator **must return** string to secure uniqueness name of locking.
```
Comment (dblock) [inline README.md]: either must return String or must return a string

File: UPGRADING.md [hunk 9]
```diff
@@ -1,4 +1,20 @@
-## Upgrading Mongoid Locker
+## Upgrading Mongoid-Locker
+
+## Upgrading to 2.0.0
+Mongoid-Locker supports only `5`, `6` and `7` versions of Mongoid.
+Since this version `Mongoid::Locker` uses unique name of locking and time is set by MongoDB. `Mongoid::Locker` no longer uses `locked_until` field and this field may be deleted. You must define new `locking_name` field of `String` type.
```
Comment (dblock) [inline UPGRADING.md]: Users should be told to call `Model.unset(:locked_until)` if they want to clean this up? Maybe add.

**Comment by dblock** [comment]
One thought, when upgrading I think users have to do some work on dropping/creating indexes? Lets make it easy for them?

File: .travis.yml [hunk 10]
```diff
@@ -5,58 +5,42 @@ language: ruby
 
 cache: bundler
 
-before_install:
-  - gem update --system
-  - gem install bundler
+script:
+  - bundle exec rspec
```
Comment (dblock) [inline .travis.yml]: I believe `rubocop` is no run in Travis. It should be.

**Comment by dks17** [comment]
Additional notes to upcoming upgrade.

Method `#with_lock` does not accept `:timeout` option anymore. We can not set specific number of seconds until the lock is considered as expired for particular document. Instead of this we can only set common `lock_timeout` for whole collection, but not for specific document. This change does not look like upgrade.

Another not implemented thing is opportunity to use `updated_at` field instead of extra `locked_at` field. There are several limitations to this. Every time when we try to get lock for a document `updated_at` field will be touched, even if we do nothing in the block of `#with_lock` method. It may effect DB indexes and caching. Mongoid provides updating the `updated_at` field with each changing of a document, and uses client time to do it. We get two providers of current time (client and MongoDB) which can lead to incorrect locking time. The opportunity is only suitable when client and MongoDB server work on the same machine, and you know that you do.

**Comment by dks17** [comment]
> One thought, when upgrading I think users have to do some work on dropping/creating indexes? Lets make it easy for them?

Added this index at the top example in `README.md`:
```ruby
class User
  include Mongoid::Document
  include Mongoid::Locker

  field :locking_name, type: String
  field :locked_at, type: Time

  field :age, type: Integer

  # optional
  # index({ _id: 1, locking_name: 1 }, { name: 'mongoid_locker_index', sparse: true, unique: true })
end
```

File: .travis.yml (hunk 10 above)
Comment (dks17) [reply .travis.yml]: Fixed.

**Comment by dks17** [comment]
> One of the lines below found in CHANGELOG.md doesn’t match the expected format. Please make it look like the other lines, pay attention to periods and spaces.

Slightly unformat entry. It's a pity.

**Comment by dblock** [comment]
> > One of the lines below found in CHANGELOG.md doesn’t match the expected format. Please make it look like the other lines, pay attention to periods and spaces.
> 
> Slightly unformat entry. It's a pity.

Make it look like all the other lines and it will be happy ;)

**Comment by dks17** [comment]
Thank you.

**Comment by joeyAghion** [comment]
Hi old friends!

When attempting an upgrade to v2.x, we discovered what I think is an undocumented behavior change, that `retries` now defaults to infinity while it used to default to `0`. This manifested as some requests (and tests) that used to be rejected starting to succeed after enough retries.

I recommend mentioning this in `UPGRADING.md`. I can PR that change unless you'd rather revert the behavior change.

**Comment by dblock** [comment]
We should treat it as a regression, and if there's no way to fix it, then document in UPGRADING. Will take either PR. Thanks!

**Pull request merged**